client.go

```diff
@@ -0,0 +1,332 @@
+	}
+}
+
+// WithExistingRootFS uses an existing root filesystem for the container
+func WithExistingRootFS(id string) NewContainerOpts {
+	return func(ctx context.Context, client *Client, c *containers.Container) error {
+		// check that the snapshot exists, if not, fail on creation
+		if _, err := client.snapshotter().Mounts(ctx, id); err != nil {
+			return err
+		}
+		c.RootFS = id
+		return nil
+	}
+}
+
+// WithNewRootFS allocates a new snapshot to be used by the container as the
+// root filesystem in read-write mode
+func WithNewRootFS(id string, image *Image) NewContainerOpts {
+	return func(ctx context.Context, client *Client, c *containers.Container) error {
+		diffIDs, err := image.i.RootFS(ctx, client.content())
+		if err != nil {
+			return err
+		}
+		if _, err := client.snapshotter().Prepare(ctx, id, identity.ChainID(diffIDs).String()); err != nil {
+			return err
+		}
+		c.RootFS = id
+		return nil
+	}
+}
+
+// WithNewReadonlyRootFS allocates a new snapshot to be used by the container as the
+// root filesystem in read-only mode
+func WithNewReadonlyRootFS(id string, image *Image) NewContainerOpts {
+	return func(ctx context.Context, client *Client, c *containers.Container) error {
+		diffIDs, err := image.i.RootFS(ctx, client.content())
+		if err != nil {
+			return err
+		}
+		if _, err := client.snapshotter().View(ctx, id, identity.ChainID(diffIDs).String()); err != nil {
+			return err
+		}
+		c.RootFS = id
+		return nil
+	}
+}
+
+func WithRuntime(name string) NewContainerOpts {
+	return func(ctx context.Context, client *Client, c *containers.Container) error {
+		c.Runtime = name
+		return nil
+	}
+}
+
+// NewContainer will create a new container in container with the provided id
+// the id must be unique within the namespace
+func (c *Client) NewContainer(ctx context.Context, id string, spec *specs.Spec, opts ...NewContainerOpts) (*Container, error) {
+	data, err := json.Marshal(spec)
+	if err != nil {
+		return nil, err
+	}
+	container := containers.Container{
+		ID:      id,
+		Runtime: c.runtime,
+		Spec: &protobuf.Any{
+			TypeUrl: specs.Version,
+			Value:   data,
+		},
+	}
+	for _, o := range opts {
+		if err := o(ctx, c, &container); err != nil {
+			return nil, err
+		}
+	}
+	r, err := c.containers().Create(ctx, &containers.CreateContainerRequest{
+		Container: container,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return containerFromProto(c, r.Container), nil
+}
+
+type PullOpts func(*Client, *PullContext) error
+
+type PullContext struct {
+	Resolver remotes.Resolver
+	Unpacker Unpacker
+}
+
+func defaultPullContext() *PullContext {
+	return &PullContext{
+		Resolver: docker.NewResolver(docker.ResolverOptions{
+			Client: http.DefaultClient,
+		}),
+	}
+}
+
+func WithPullUnpack(client *Client, c *PullContext) error {
+	c.Unpacker = &snapshotUnpacker{
+		store:       client.content(),
+		diff:        client.diff(),
+		snapshotter: client.snapshotter(),
+	}
+	return nil
+}
+
+type Unpacker interface {
```

Comment:
@crosbymichael The problem is that, we expect `Image` and `Container` also to be an interface. It will be harder to wrap them all on our side, e.g. https://play.golang.org/p/JvlZhLONen.